presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java

```diff
@@ -261,7 +261,7 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                 new IterativeOptimizer(
```

Comment:
`Later, when AddExchanges.planTableScan executes...`

After this change, in case of evil expression, which uses partitioning column and does not prune any partition, this problem will be even bigger.  Because now `PickTableLayout` will also build an expression out of the current constraint... in general it will also execute `shouldPrune`. 

Notice `PickTableLayout` in normal (happy path) case it executed 4 times. With and without predicate and times 2 to make `IterativeOptimizer` stop the iteration. `PickTableLayout` is additionally used twice in a list optimizers so it makes it even bigger problem ; /

Reply:
Fair enough. The underlying problem is that interpreting those expressions is *super* expensive. Without a revamp of the interpreter and the IR, there's no way to make the performance of individual evaluations faster.

I'm trying out another approach, which I'll post soon. Closing this one for now.